### ISSUE TITLE
Implement 'incremental' provision support

### ISSUE THREAD

**Issue by SubPointSupport** [SubPointSolutions/spmeta2#553]
### Work in progress
* [x] Implement DefaultIncrementalModelTreeTraverseService
* [x] Regression tests: basic API
* [x] Regression tests: all definitions + double deployment
* [x] Regression tests: all definitions + double deployment and random updates
* [x] Implement special provision services (both CSOM/SSOM) 
* [x] Implement fluent API to make existing provision services working with incremental updates
* [x] Implement fluent API totoggle provision services mode - from defalt to incremental
* [x] Regression tests for newly created provision services and fluent API
* [x] Implement IncrementalModelPrettyPrintService (to output model state)
* [x] model hash classes / api
* [x] Persistence storage base service
* [x] Persistence storage 'file system' implementation, tests
* [x] Persistence storage CSOM - DefaultCSOMWebPropertyBagStorage
* [x] Persistence storage SSOM - DefaultSSOMWebPropertyBagStorage
* [x] Persistence storage SSOM - DefaultSSOMWebApplicationPropertyBagStorage
* [x] Persistence storage SSOM - DefaultSSOMFarmPropertyBagStorage
* [x] Add option for persistence storage auto-detection

### Problem
SPMeta2 might take unreasonable amount of time on large deployments. For instance, thousands of definitions and/or terabytes of content in the farm cause provision time to be unreasonable huge.

### Solution 
SPMeta2 needs to offer a concept of "incremental" or "conditional" provision. In nutshell, SPMeta2 should be able to calculate changes and make a derision on "should/should not" provision the current artifact.

Benefits of incremental provision should enable robust, fast provision on extremely large SharePoint deployments - terabytes of data and thousands of artifacts.

Potential implementation should work smoothly with already existing code, legacy code and current deployments. We should not spend tons of time adding some code or troubleshooting stuff.

### Implementation
Two potential implementation to calculate differences are seen:
* Model based - difference between current deployed model and previously deployed model
* SharePoint based - difference between current deployed model  and actual SharePoint objects

To simplify things, the following terminology is used:

* "model hash" - represents model hash (model + model nodes + definitions)
* "node hash"  - represents model node hash (node + definition)
* "definition hash"  - represents model node hash (node + definition)

With two strategies to calculate the difference, the easiest way would be to implement diff calculation between models. Calculations between model and SharePoint object is complex and require a huge amount of code to be written for both CSOM, SSOM and O365 APIs.

First draft of the "model based" diff calculation should work as following:
* calculate model hash of the current model
* fetch model hash of previously deployed model (via extensible provider: from local folder, sharepoint or other storage)
* calculate the diff between models and mark nodes as "should provision" or "should not provision"
* provision model
* provide hash of both current and previous model

A set of additional services are to be provided:
* a service to calculate model hashes
* a service to persist model hash (ootb impl for folders, SharePoint CSOM, SharePoint SSOM)
* a service to output "final model" with marked nodes [will / will not be provisioned]

Suggested implementation has to be implemented for both CSOM/SSOM APIs, covered by unit tests and regression tests, offer traceability to understand which nodes are/aren't provisioned and why.

Benefits of incremental provision should enable robust, fast provision on extremely large SharePoint deployments - terabytes of data and thousands of artifacts.

**Comment by andreasblueher** [comment]
Hey, since you added the plugin label I suppose this feature won't come soon right? I'm working on a project with 300 custom fields, several lists and content types and provisioning everything in a single feature activation is simply no longer possible, because we would run into timeouts more often. We split up the provisioning to reduce execution time, but in the end it's just a workaround.

**Comment by SubPointSupport** [comment]
Totally understandable, supporting similar project with same-same amount of fields, content types and the rest of the artefacts. 

It can be recommended splitting up models into a smaller, "sub-model" by the functionality, sprint, version or other logic which makes sense for you project. Not only to improve provision time, but also improve maintability and supportability of the whole project. Once there smaller models, there might be a good idea to optionally turn on/turn off provision via boll-flags or configuration values. Something like that:

* ShouldDeployHomeSiteFieldAndContentTypes
* ShouldDeployStyleLibrary
* ShouldDeploySandboxSolutions
* ShouldDeployThat
* ShouldDeployThis
* ShouldDeployUpdatesVersion2
* ShouldDeployUpdatesVersion3
* so on..

As for the incremental updates/provision, we still thinking about possible implementations along with the feedback. Not many people asked for it, so we feel quite comfortable to delay that once and push others. "Plugins" label is made only for better tracking at git. All these tickets will be implemented within M2 library, nothing is changed.

**Comment by SubPointSupport** [comment]
Obsolete. Will be implemented within M2 v1.3.xx

**Comment by rolandoldengarm** [comment]
I would like to see this feature. On a dev environment without a lot of documents, it takes about 30 minutes to deploy a model with ~20 content types and ~40 fields. 
On prod we have a significant amount of data, and I'm afraid it will take forever there.

**Comment by SubPointSupport** [comment]
Thanks @rolandoldengarm, we put together more detailed specs in the ticket description.

If you and @andreasblueher could please have a look to ensure we are on the right path? We see a great and easy implementation here which can be delivered in several days. Besides, we really want to switch to 1.3.x.x versioning, kinda tired with 1.2.x.x :)

**Comment by rolandoldengarm** [comment]
Yeah awesome, looks good to me. Key thing is that the model definition in C# may depend on sp artefacts like js/css files, or farm solutions. Means that even though the model definition didn't change, the hash could change.

**Comment by SubPointSupport** [comment]
True. But don't forget that all the content is stored in actual definition as byte arrays. Meaning, definition contains all the information on how it needs to be provisioned including the actual content of the files, solutions and so on. 

Actual hash of a definition is a hash of its serialized representation. That's how all the changes in files and solutions are propagated into definition, its serialized content and hash at the end.

Nevertheless, not only it should be fine in theory but it will have to be covered by unit/regression tests.

**Comment by rolandoldengarm** [comment]
:+1:

**Comment by ReDERO** [comment]
Excellent idea! As I see M2 team in v1.3 will bring to us many interesting things, like parallel provisioning and incremental update. It's really nice features.

We have the same situation as @rolandoldengarm desribes previously: to dev environment deploying process takes near 30 minutes. And when we get an error on the middle of the process, it`s very painful to restart your deploy with fixes, because you understand that the half time you deploy design files, that didn't affect on provision. So to make provision process faster we need to comment some line of codes in web model and it's not very good.

IMHO, incremental update is a required key to the provision universe and will be a good feature for SPMeta2)

**Comment by SubPointSupport** [comment]
Okay, first beta implementation looks as following:
```cs
// create you service as usual
var provisionService = ...;

// set provision model via extension method
provisionService.SetIncrementalProvisionMode();
// set previous model hash if you've got it (see below on how to obtain it)
provisionService.SetIncrementalProvisionModelHash(previousModelHash);

// deploy model as usual
provisionService.DeployModel()

// this is the model hash to be saved somewhere, it stores information on provisioned model
var modelHash = provisionService.GetIncrementalProvisionModelHash();

// optionally, switch back to default provision
provisionService.SetDefaultProvisionMode();
```

Once model is deployed, we get two things such as model hash and some userful trace:
```cs
// print model state
var tracer = new DefaultIncrementalModelPrettyPrintService();

Console.WriteLine(string.Format("Deployed model with incremental updates:"));
Console.WriteLine(Environment.NewLine + tracer.PrintModel(model));
```

Outcome with first provision:
```
[-] [WebDefinition] - Title:[] Description:[] Convert(p.LCID):[1033]
[+]     [ListDefinition] - Title: [662fecf933624f81934bef9a18e6343b] 
[+]         [FolderDefinition] - Name:[ce56514be250472fa8d44c9aede36fda]
[+]         [FolderDefinition] - Name:[c416cebc8f8b4f2abe8491102d0108ee]
[+]         [ListItemDefinition] - Title:[04c00dbf66ea4d2fbb23d32fae87d07f]
```

Outcome with second and further provision:
```
[-] [WebDefinition] - Title:[] Description:[] Convert(p.LCID):[1033]
[-]     [ListDefinition] - Title: [662fecf933624f81934bef9a18e6343b] 
[-]         [FolderDefinition] - Name:[ce56514be250472fa8d44c9aede36fda]
[-]         [FolderDefinition] - Name:[c416cebc8f8b4f2abe8491102d0108ee]
[-]         [ListItemDefinition] - Title:[04c00dbf66ea4d2fbb23d32fae87d07f]
```

Plus indicates provision and minus - skipping actual provision of the artifact. 

Most of the ground work has been done, we already re-engineered out regression testing to support both normal and incremental provision. Working hard on two things right now:
* More regression testing, adding tests on randomized changes and incremental provision checks
* Implementing default persistence storage for model hash to work with local file system AND SharePoint

Ideally, you wouldn't have to work with model hash directly as it would always be automatically persist into SharePoint. You would have to call .SetIncrementalProvisionMode() and then also set "unique ID"  of the model so that we would be able to identify the model within many model hashes.

Looking forward to your ideas and feedback on the current progress. We are moving really fast with that feature aiming to push it as beta release early next week.

**Comment by rolandoldengarm** [comment]
Wow you guys are amazing @SubPointSupport  👍  Looking forward to implementation.

**Comment by SubPointSupport** [comment]
Implemented. Testing but beta access can be done as following:

VS2012+ NuGet feed:
https://www.myget.org/F/subpointsolutions-staging/api/v2

VS2012+ NuGet feed:
https://www.myget.org/F/subpointsolutions-staging/api/v3/index.json

Install-Package -Id "package-name" -Version 1.2.120-beta1

Both CSOM/SSOM provision with auto-detection for model hash persistence:
```cs
var incrementalProvisionConfig = new IncrementalProvisionConfig();

// in that case M2 auto-detect runtime/api and stored model hash in Property Bag at web, web app or farm level after successful provision
incrementalProvisionConfig.AutoDetectSharePointPersistenceStorage = true;

// configure incremental provision as per options
provisionService.SetIncrementalProvisionMode(incrementalProvisionConfig);

// assign ID, model should have that ID so that SPMeta2 would be able to find/save model hash using that key
// use "my-model", "my.model" or something like that
model.SetIncrementalProvisionModelId(incrementalModelId);

// deploy model as usual
 provisionService.DeployModel(host, model)

// get back to normal provision
provisionService.SetDefaultProvisionMode();
```

Trace WHAT actually was or was NOT provisioned:
```cs

provisionService.DeployModel(host, model)

var tracer = new DefaultIncrementalModelPrettyPrintService();

Console.WriteLine(string.Format("Deployed model with incremental updates:"));
Console.WriteLine(Environment.NewLine + tracer.PrintModel(model));

```

Trace in real time:

```cs

provisionService.OnModelNodeProcessed += (sender, args) =>
            {
               var incrementalRequireSelfProcessingValue = modelNode.NonPersistentPropertyBag
                .FirstOrDefault(p => p.Name == "_sys.IncrementalRequireSelfProcessingValue");

            if (incrementalRequireSelfProcessingValue != null)
                shouldDeploy = ConvertUtils.ToBoolWithDefault(incrementalRequireSelfProcessingValue.Value, false);
        
               var wasDeployed = incrementalRequireSelfProcessingValue.Value;

                Trace.WriteLine(
                    string.Format("[5] Processed: [{0}/{1}] - [{2}%] - [{3}] [{4}]",
                    new object[] {
                                  args.ProcessedModelNodeCount,
                                  args.TotalModelNodeCount,
                                  100d * (double)args.ProcessedModelNodeCount / (double)args.TotalModelNodeCount,
                                  args.CurrentNode.Value.GetType().Name,
                                  args.CurrentNode.Value,
                                  wasDeployed == true ? "[+]" : "[-]"}));
            };

provisionService.DeployModel(host, model)
```

Manual management for model hash (so that you can take care of storing it somewhere:
```cs
provisionService.SetIncrementalProvisionMode();

// set model hash if you have one
provisionService.SetIncrementalProvisionModelHash( hash );

// provision as usual
provisionService.DeplyModel( host, model); 

// get new model hash, save it somewhere and then use in .SetIncrementalProvisionModelHash() method
var newMOdelHost = provisionService.GetIncrementalProvisionModelHash();
```

Using file system to save model hash:
```cs
var incrementalProvisionConfig = new IncrementalProvisionConfig();
// use overload of DefaultFileSystemPersistenceStorage(folderPath) to get state in your folder
// default folder is Environment.SpecialFolder.LocalApplicationData + SPMeta2
incrementalProvisionConfig.PersistenceStorages.Add(new DefaultFileSystemPersistenceStorage());

provisionService.SetIncrementalProvisionMode(incrementalProvisionConfig);

// set model id for incremental provision
model.SetIncrementalProvisionModelId(incrementalModelId);

// deploy as usual
provisionService.DeployModel(host, model)
```

Have fun, we'll be testing more over the weekend maturing it for the next week. Docs will be updated accordingly.

**Comment by rolandoldengarm** [comment]
Thanks, will implement it next week 👍

**Comment by rolandoldengarm** [comment]
It works amazing. One deployment locally crashed half way, and then next run resumed where it crashed before.
Previously we worked around this by commenting out code, but that is hacky.

This is a must have for all SPMeta2 developers.

Two comment about the example code:

1. Maybe this can be implemented as getter on the ModelNode class?
````
var incrementalRequireSelfProcessingValue = modelNode.NonPersistentPropertyBag
                .FirstOrDefault(p => p.Name == "_sys.IncrementalRequireSelfProcessingValue");
````
2. The shouldDeploy/wasDeployed code did not compile. I've changed it to:

````
var incrementalRequireSelfProcessingValue = args.CurrentNode.NonPersistentPropertyBag
                .FirstOrDefault(p => p.Name == "_sys.IncrementalRequireSelfProcessingValue");
            bool shouldDeploy = true;
            if (incrementalRequireSelfProcessingValue != null)
            {
                shouldDeploy = ConvertUtils.ToBoolWithDefault(incrementalRequireSelfProcessingValue.Value, false);
            }
````

Awesome stuff @SubPointSupport , thanks for the super quick turnaround on this one. Hopefully it can be pushed to NuGet soon :)

**Comment by SubPointSupport** [comment]
Thanks, @rolandoldengarm!

>Maybe this can be implemented as getter on the ModelNode class?
Yes, it would be wrapped into extension method similar to .SetIncrementalProvisionMode() and other extensions.

>The shouldDeploy/wasDeployed code did not compile. I've changed it to:
Correct, our bad while copy-pasting stuff here in the ticket.

1.2.120-beta1 is already in NuGet. We are running some test preparing for the major release early next week. Docs and guides will be updated as well.

**Comment by andreasblueher** [comment]
Hey guys, 

haven't had the chance to look into this but with close to 1000 models being deployed in some of our customer solutions I'm more than happy about how fast things went here.

I'm not sure I'll be able to check it out without the next 2 or 3 weeks, but since we're planning on upgrading to the latest SPMeta version soon (currently on 1.2.60) this will definitely be part of my tests.

**Comment by SubPointSupport** [comment]
1000 models? Wow, impressive.

Incremental provision is disabled by default. Full backward compatibility over all SPMeta2 version upgrades. You'll have to enable it manually, with your code.

**Comment by andreasblueher** [comment]
This change has cut feature activation time for one of my solutions in half! This is a huge relieve and allows us to deploy more easily. Thank you!

**Comment by SubPointSupport** [comment]
Sweet! Could you give us an overall stat? What are the artifacts, how many of them, and what's the time benefits? 

We'll be improving this feature over the following weeks introducing a planning feature and better diff calculation. That way, with the new calculation, a diff of the mode would be calculated BEFORE deployment which would cut the deployment time even further. Appx change is around 10-50 times.

**Comment by andreasblueher** [comment]
The solution where I tested #1005 and #553 had ~500 definitions, mostly FieldDefinitions but also ListDefinitions, UserCustomActionDefinitions, ContentTypeDefinitions. Activating the feature initially took me about 60 seconds. I had verbose logging activated so maybe it could have been faster without it.

Your ideas about pre calculating sounds great and I would love to give you more feedback on other solutions where SPMeta is being used (1000 definitions easily) but I would need you to ship 1.2.130.

**Comment by andreasblueher** [comment]
Here some other test results:

Deploying 189 FieldDefinition through a SharePoint feature:
Standard: 16s
Incrementell: 1,5s
----------------
Deploying 310 contenttypes, contenttypelinks, webfeatures and security group definitions:
Standard: 98s
Incrementell: 92s
ContentTypeFieldLinkDefinitions appeared to be very slow compared to fielddefinitions especially. I would argue that here was no real improvement at all compared to the first run.
----------------
Deploying 125 lists, listviews, module files, webparts:
Standard: 18s
Incrementell: 7s

2 out of 3 show a very good improvement (especially the first example), but maybe you can look into contenttypefieldlink definitions again and why they could perform so badly.

**Comment by SubPointSupport** [comment]
This is pure gold. Thank you for the feedback, we are really excited to see this in action for you.

We are aware that ContentTypeFieldLinkDefinitions aren't playing well yet, was raised by other people as well. Looked into it, couldn't spot anything. Going to get back on it, rework and try to test better.

**Comment by SubPointSupport** [comment]
We bet that it would be all about pre-calculating the model before actual provision.
> Your ideas about pre calculating sounds great and I would love to give you more feedback on other solutions where SPMeta is being used (1000 definitions easily) but I would need you to ship 1.2.130.

The thing is that currently incremental provision still fetches artefacts from SharePoint skipping the update logic. It's not too bad - we don't change things we don't have to but this is only a half of the solution, we need to coved the second half and cut artefact fetching logic. That way, SPMeta2 would skip a fare share of `WithResolvingModelHost` calls which, most likely, cause these time drifts for different models/artefacts.

**Comment by koltyakov** [comment]
Hey guys @SubPointSupport, first off thanks a lot for the feature!
It becomes vital on some projects with really huge models, there it's difficult to separate the models due to organizational and global team related nuances.
Just wanted to add our five cents that improving incrementally provisioning by excluding redundant data fetching is in enormous demand.
If you can put this enhancement in your product plans we'll be super happy to use it.

**Comment by SubPointSupport** [comment]
Not a problem, thanks for raising this again, @koltyakov 

Out of curiosity, what are the challenges with the current performance and your project related workflow? Pre-calculating IS on the radar, we might do that really-really soon, curious on your vision  of the situation as well.

**Comment by koltyakov** [comment]
Well, we're are in a transition and adoption stage with a remote team. Where existing project dictates some rules about how to organize and manage the artifacts. The project is rather big. A model for a web (an area for a teamwork, a number of such areas can be created constantly) includes more than 3000 artifacts and this number grows. The reasons are objective, cause the app is really sophisticated.

In this process, we're helping a remote team over the ocean to master M2 which is really a relief and a huge step forward in terms of improving and managing artifacts in the mentioned project.

Also during the transition period, some team members got to deploy from their machines to the servers located on the different continent (just because there is no other choice for now). The process of deployment takes time. Sometimes it can fail due to network issues, then the process starts from the beginning.

It's too early to chunk the models to small parts all the time as it can bring extra complexity for the members who are still new to M2, and can create hidden issues (e.g. something was added to a small model, but had been forgotten to be added to a full solution model).

In parallel, the work is in progress to push some deployment to CI/CD tools and to deploy within the machines closer to the servers.

Also, a solution itself assumes definitions update. Let's say, 100 webs had been created with one version, then a new release should be delivered to all of the existing webs.

In other words, speeding up the second run (incremental) delivery in this scenario is a really huge deal for us and a reason for appreciation.

**Comment by SubPointSupport** [comment]
Alright, a few points to improve and/or create separate tickets.

> Sometimes it can fail due to network issues, then the process starts from the beginning.
SPMeta2 has a built-in re-try logic for CSOM. It picks up disconnections, waits, re-tries. There is a pluggable API to extend out of the box handlers up for your scenario so that custom offline/network issues can be handled by you.

> It's too early to chunk the models to small parts all the time as it can bring extra complexity for the members who are still new to M2, and can create hidden issues (e.g. something was added to a small model, but had been forgotten to be added to a full solution model).

Totally agree.

> In other words, speeding up the second run (incremental) delivery in this scenario is a really huge deal for us and a reason for appreciation.

Agree again, totally understood. Let us see what can be done here. It isn't a major change or effort but it required some internal knowledge and context. Might push somewhat next week.

**Comment by koltyakov** [comment]
Thank you in advance! You guys are the best!

**Comment by rolandoldengarm** [comment]
On a very large environment (like 30 site collections, 4TB of data, millions of documents), even with incremental support, it just took too long. We've implemented a migration styled framework using https://github.com/jackawatts/ionfar-sharepoint-migration/tree/master/src

This means there is an initial migration to deploy the entire model, and migrations to deploy changes (like adding fields).
This means a bit more thinking / testing, but it's very fast.

**Comment by SubPointSupport** [comment]
@rolandoldengarm, some context around this:

SPMeta2 still fetches artefacts over incremental provision. While it's still much more faster, it comes with a drawback. Imagine we've got a content type with 5 content field links, and only 2 content type fields links were changed. The way SPMeta2 works is that it treats each content type field link as a separate entity. Here is the current flow which would be called 5 times:
- resolve content type
- update content type field link
- update content type (push changes to children)

With incremental provision, this flow would be still called twice for only changed content type field links. And, potentially, three times for non-changed field links with skipping updates BUT still, most likely, updating the content type and pushing changes to the lists. 

Totally on SPMeta2, some internal stuff and historical design on how it processes and handles artefacts. Several improvements we can do:

* Pre-calculate "model diff" before deployment (which we already looking into) 
That would calculate the real diff of the model so it would save lookup logic and IDLE artefact resolutions.

* Improve content type provision - don't trigger SPContentType.Update(true) if no changes on the content type field link level were detected (applicable only for the incremental provision)

* Batch content type field link changes (and, potentially, list field links - any other "links" artefacts) to save artefact resolutions and updates - resolve artefact once, batch all children content type field links into one change, and then call SPContentType.Update(true) once opposite, per batch, rather than per every content type field link 

Such batching would improve both regular and incremental provision. Requires some drastic internal re-engineering due to the flow on how SPMeta2 handles parent-child artefact provision. It does not group artefacts, it always sees pair of "parent-child" splitting all children into separate "parent-child" pairs. For most of the artefacts that's fine, but Content Type flow works slightly different in SharePoint - hence all challenges we see.

All things mentioned here have an impact on what @koltyakov and other guys in the thread have experienced. Pointless artefact resolution, pointless artefact update (while resolving parent for the child) and, finally, batching over "groupable" children such as content type field links.

Hope we can solve that puzzle soon and if so, that would be one of the killer feature we ever had over the last 3-4 years.
